### PR TITLE
MultiValue VarByte V4 index writer and consolidate V4 reader for all types

### PR DESCRIPTION
**Pull request by saurabhd336**
Adding support for serializing and writing MV columns in `VarByteChunkForwardIndexWriterV4` and adding new V4 reader that can be used to read SV var length, MV fixed length and MV var length buffers encoded with V4 writer.

### CODE REVIEW

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/writer/impl/VarByteChunkWriter.java [hunk 1]
```diff
@@ -28,4 +28,8 @@ public interface VarByteChunkWriter extends Closeable {
   void putString(String value);
 
   void putBytes(byte[] value);
+
+  void putStrings(String[] values);
+
+  void putByteArrays(byte[][] values);
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/writer/impl/VarByteChunkWriter.java:34]: Suggest renaming them to be more clear
```suggestion
  void putStringMV(String[] values);

  void putBytesMV(byte[][] values);
```

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/writer/impl/VarByteChunkForwardIndexWriterV4.java [hunk 2]
```diff
@@ -121,6 +123,42 @@ public void putBigDecimal(BigDecimal bigDecimal) {
     putBytes(BigDecimalUtils.serialize(bigDecimal));
   }
 
+  @Override
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/writer/impl/VarByteChunkForwardIndexWriterV4.java:126]: (minor) Suggest keeping the method the same order as in the interface to be easier to navigate

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/writer/impl/VarByteChunkForwardIndexWriterV4.java [hunk 3]
```diff
@@ -121,6 +123,42 @@ public void putBigDecimal(BigDecimal bigDecimal) {
...
+    // num values + length of each value
+    int size = Integer.BYTES + Integer.BYTES * values.length;
+    for (byte[] value : values) {
+      size += value.length;
+    }
+    byte[] serializedBytes = new byte[size];
+    ByteBuffer byteBuffer = ByteBuffer.wrap(serializedBytes);
+    byteBuffer.putInt(values.length);
+
+    for (byte[] value : values) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/writer/impl/VarByteChunkForwardIndexWriterV4.java:137]: Consider using the same format as the old index `[numValues, length1, length2, ..., bytes1, bytes2, ...]`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/forward/BaseVarByteChunkForwardIndexReaderV4.java [hunk 4]
```diff
@@ -43,9 +43,9 @@
  * (BIG_DECIMAL, STRING, BYTES).
  * <p>For data layout, please refer to the documentation for {@link VarByteChunkForwardIndexWriterV4}
  */
-public class VarByteChunkSVForwardIndexReaderV4
-    implements ForwardIndexReader<VarByteChunkSVForwardIndexReaderV4.ReaderContext> {
-  private static final Logger LOGGER = LoggerFactory.getLogger(VarByteChunkSVForwardIndexReaderV4.class);
+public class BaseVarByteChunkForwardIndexReaderV4
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/forward/BaseVarByteChunkForwardIndexReaderV4.java:46]: Rename it to `VarByteChunkForwardIndexReaderV4`. Also update the javadoc for this class

**Review by Jackie-Jiang — COMMENTED** [review]
Mostly good

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/writer/impl/VarByteChunkForwardIndexWriterV4.java (hunk 3 above)
Comment (saurabhd336) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/writer/impl/VarByteChunkForwardIndexWriterV4.java:137]: Ack

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/forward/BaseVarByteChunkForwardIndexReaderV4.java (hunk 4 above)
Comment (saurabhd336) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/forward/BaseVarByteChunkForwardIndexReaderV4.java:46]: Ack

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/writer/impl/VarByteChunkWriter.java (hunk 1 above)
Comment (saurabhd336) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/writer/impl/VarByteChunkWriter.java:34]: Ack

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/forward/ForwardIndexReaderFactory.java [hunk 5]
```diff
@@ -81,13 +81,17 @@ public static ForwardIndexReader createRawIndexReader(PinotDataBuffer dataBuffer
             ? new FixedBytePower2ChunkSVForwardIndexReader(dataBuffer, storedType)
             : new FixedByteChunkSVForwardIndexReader(dataBuffer, storedType);
       } else {
-        return version == VarByteChunkForwardIndexWriterV4.VERSION ? new VarByteChunkSVForwardIndexReaderV4(dataBuffer,
-            storedType) : new VarByteChunkSVForwardIndexReader(dataBuffer, storedType);
+        return version == VarByteChunkForwardIndexWriterV4.VERSION ? new VarByteChunkForwardIndexReaderV4(
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/forward/ForwardIndexReaderFactory.java:84]: Consider reversing the check to first check the version

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/forward/VarByteChunkForwardIndexReaderV4.java [hunk 6]
```diff
@@ -39,13 +39,13 @@
 
 
 /**
- * Chunk-based single-value raw (non-dictionary-encoded) forward index reader for values of variable length data type
- * (BIG_DECIMAL, STRING, BYTES).
+ * Chunk-based single-value raw (non-dictionary-encoded) forward index reader for values of SV variable length data type
+ * (BIG_DECIMAL, STRING, BYTES), MV fixed length and MV variable length data types.
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/forward/VarByteChunkForwardIndexReaderV4.java:43]:
```suggestion
 * Chunk-based raw (non-dictionary-encoded) forward index reader for values of SV variable length data type
 * (BIG_DECIMAL, STRING, BYTES), MV fixed length and MV variable length data types.
```

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/forward/VarByteChunkForwardIndexReaderV4.java [hunk 7]
```diff
@@ -113,6 +116,156 @@ public byte[] getBytes(int docId, ReaderContext context) {
...
+    ByteBuffer byteBuffer = ByteBuffer.wrap(context.getValue(docId));
+    int numValues = byteBuffer.getInt();
+    int contentOffset = (numValues + 1) * Integer.BYTES;
+    for (int i = 0; i < numValues; i++) {
+      int length = byteBuffer.getInt((i + 1) * Integer.BYTES);
+      byte[] bytes = new byte[length];
+      byteBuffer.position(contentOffset);
+      byteBuffer.get(bytes, 0, length);
+      valueBuffer[i] = new String(bytes, StandardCharsets.UTF_8);
+      contentOffset += length;
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/forward/VarByteChunkForwardIndexReaderV4.java:214]: Do we need to maintain contentOffset? `position` it once before the for loop should be good

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/forward/VarByteChunkForwardIndexReaderV4.java [hunk 8]
```diff
@@ -113,6 +116,156 @@ public byte[] getBytes(int docId, ReaderContext context) {
...
+  @Override
+  public int getStringMV(int docId, String[] valueBuffer, VarByteChunkForwardIndexReaderV4.ReaderContext context) {
+    ByteBuffer byteBuffer = ByteBuffer.wrap(context.getValue(docId));
+    int numValues = byteBuffer.getInt();
+    int contentOffset = (numValues + 1) * Integer.BYTES;
+    for (int i = 0; i < numValues; i++) {
+      int length = byteBuffer.getInt((i + 1) * Integer.BYTES);
+      byte[] bytes = new byte[length];
+      byteBuffer.position(contentOffset);
+      byteBuffer.get(bytes, 0, length);
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/forward/VarByteChunkForwardIndexReaderV4.java:212]:
```suggestion
      byteBuffer.get(bytes);
```

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/forward/VarByteChunkForwardIndexReaderV4.java [hunk 9]
```diff
@@ -113,6 +116,156 @@ public byte[] getBytes(int docId, ReaderContext context) {
...
+      valueBuffer[i] = new String(bytes, StandardCharsets.UTF_8);
+      contentOffset += length;
+    }
+    return numValues;
+  }
+
+  @Override
+  public String[] getStringMV(int docId, VarByteChunkForwardIndexReaderV4.ReaderContext context) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/forward/VarByteChunkForwardIndexReaderV4.java:220]: Same for these methods

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/writer/impl/VarByteChunkForwardIndexWriterV4.java [hunk 10]
```diff
@@ -142,6 +144,51 @@ public void putBytes(byte[] bytes) {
...
+    ByteBuffer byteBuffer = ByteBuffer.wrap(serializedBytes);
+    byteBuffer.putInt(values.length);
+    byteBuffer.position(headerSize);
+    for (int i = 0; i < values.length; i++) {
+      byte[] utf8 = values[i].getBytes(UTF_8);
+      byteBuffer.putInt((i + 1) * Integer.BYTES, utf8.length);
+      byteBuffer.put(utf8);
+    }
+
+    putBytes(byteBuffer.array());
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/writer/impl/VarByteChunkForwardIndexWriterV4.java:167]:
```suggestion
    putBytes(serializedBytes);
```

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/writer/impl/VarByteChunkForwardIndexWriterV4.java [hunk 11]
```diff
@@ -142,6 +144,51 @@ public void putBytes(byte[] bytes) {
...
+    byte[] serializedBytes = new byte[size];
+    ByteBuffer byteBuffer = ByteBuffer.wrap(serializedBytes);
+    byteBuffer.putInt(values.length);
+    byteBuffer.position(headerSize);
+    for (int i = 0; i < values.length; i++) {
+      byteBuffer.putInt((i + 1) * Integer.BYTES, values[i].length);
+      byteBuffer.put(values[i]);
+    }
+
+    putBytes(byteBuffer.array());
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/writer/impl/VarByteChunkForwardIndexWriterV4.java:189]:
```suggestion
    putBytes(serializedBytes);
```

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise. Nice work!

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/writer/impl/VarByteChunkForwardIndexWriterV4.java (hunk 10 above)
Comment (saurabhd336) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/writer/impl/VarByteChunkForwardIndexWriterV4.java:167]: Ack

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/forward/VarByteChunkForwardIndexReaderV4.java (hunk 8 above)
Comment (saurabhd336) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/forward/VarByteChunkForwardIndexReaderV4.java:212]: Ack

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/forward/VarByteChunkForwardIndexReaderV4.java (hunk 9 above)
Comment (saurabhd336) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/forward/VarByteChunkForwardIndexReaderV4.java:220]: Ack

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/forward/ForwardIndexReaderFactory.java (hunk 5 above)
Comment (saurabhd336) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/forward/ForwardIndexReaderFactory.java:84]: This logic is quite convoluted. I've updated to make it shorter, please have a look. We may think about refactoring VarByteChunkSVForwardIndexReader and VarByteChunkMVForwardIndexReader to be a single class similar to the new v4 reader.

**Pull request merged**